### DISCUSSION TITLE
High latency when serving bigger size payload

### DISCUSSION THREAD

**Discussion by anup-krai** [Kong/kong#11109] [Help]
I am trying to proxy an API through Kong which takes request payloads which are bigger in size (approx 1 GB or little less). 

I see Kong latency increases a lot when payload size increases. For 1 GB of request payload it goes from seconds to mins.

When I change [konghq.com/request-buffering](https://docs.konghq.com/kubernetes-ingress-controller/latest/references/annotations/#konghqcomrequest-buffering) to false, then Kong proxy latency comes down to ms but in this case upstream latency goes high.

When I hit directly to upstream API without Kong, latency is around 100ms.
When I hit through Kong with request-buffering set to true then Kong latency is high and upstream latency is less so total latency is going till mins.
When I hit through Kong with request-buffering set to false, Kong latency is less but upstream latency is going high so total latency is going till mins. 

Why upstream latency is going high when hit from Kong with request-buffering set to false?

How can I reduce the total latency for bigger size payloads?

I am using Kong 2.7.1, KIC 2.2.1 with Postgres.

**Comment by Manas1820** [comment]
Is this resolved ?

**Reply by anup-krai** [reply]
@Manas1820  - This issue is still not resolved.

**Comment by ProBrian** [comment]
Hi @anup-krai , Can you provide more detailed info, e.g kong configuraitons, enabled plugins, request header/body, etc.

**Reply by anup-krai** [reply]
Hi @ProBrian ,

Below are the details -

Kong - 3.7.1 with Postgres DB

When I hit Kong with a payload size of 1.7 MB I see below latency -

**Case 1** -
`route.request_buffering: true,
route.response_buffering: true,
request.size: 1704244
 "latencies": {
    "receive": 1,
    "proxy": 55,
    "request": 2402,
    "kong": 2346
  }`

**Case 2** -
`route.request_buffering: false,
route.response_buffering: true,
request.size: 1704244,
"latencies": {
	"receive": 0,
	"request": 1130,
	"kong": 1,
	"proxy": 1129
},`

When request_buffering is enabled then Kong latency is high and If I disable request_buffering then latency is moved to upstream and total latency remains almost same.
If I directly hit upstream with the same payload then its latency is hardly 10 ms.

There are no plugins added expect for the logging, metrics one. I am using default values for request size, buffer size in kong.conf file.

I have attached [request.json](https://github.com/user-attachments/files/17364820/request.json) which I am using. Upstream is any API which just returns HTTP 404 without doing processing.

**Reply by Oyami-Srk** [reply]
Hi, @anup-krai, Could you please provide more details about your deployment? Like which logging and metrics plugins were used. I'm trying to reproduce that but not lucky with that. Could you use [Request Debugger](https://docs.konghq.com/gateway/latest/production/debug-request/) to inspect the latency introduced by Kong?
